pkg/tsdb/azuremonitor/aztokenprovider/token_provider.go

```diff
@@ -180,6 +215,119 @@ func (c *clientSecretTokenRetriever) GetAccessToken(ctx context.Context, scopes
 	return &AccessToken{Token: accessToken.Token, ExpiresOn: accessToken.ExpiresOn}, nil
 }
 
+type userIdentityTokenRetriever struct {
+	tokenEndpoint string // token endpoint to retrieve the token from
+	authHeader    string // exact Authorization header to be used to talk to the token endpoint
+	client        *http.Client
+}
+
+func (c *userIdentityTokenRetriever) GetCacheKey() string {
+	return fmt.Sprintf("azure|useridtoken|%s", c.tokenEndpoint)
+}
+
+func (c *userIdentityTokenRetriever) Init() error {
+	c.client = &http.Client{Timeout: time.Second * 10}
+	return nil
+}
+
+func (c *userIdentityTokenRetriever) GetAccessToken(ctx context.Context, scopes []string) (*AccessToken, error) {
+	value := ctx.Value(proxyutil.ContextKeyLoginUser{})
+	if value == nil {
+		return nil, fmt.Errorf("failed to get signed-in user")
+	}
+
+	userId := value.(string)
+	if userId == "" {
+		return nil, fmt.Errorf("empty signed-in userId")
+	}
+
+	return c.getUserAccessToken(userId, scopes)
+}
+
+func (c *userIdentityTokenRetriever) getUserAccessToken(userId string, scopes []string) (*AccessToken, error) {
```

Comment:
Could we move the actual token client implementation out of the token provider?
To another file and struct (e.g. `userIdentityClient` etc.)